lib/shrine/storage/google_cloud_storage.rb

```diff
@@ -138,6 +138,29 @@ def copyable?(io)
         # TODO: add a check for the credentials
       end
 
+      # Google cloud storage client only accepts IO|IOString|Tempfile instances
+      # or else it will raise a "Google::Apis::ClientError, 'Invalid upload source"
+      #
+      # We need to convert our file to an IO.
+      j
+      # see:
+      # https://github.com/google/google-api-ruby-client/blob/1c2cf5d57fd5e606c03f2aecab88469f46b8f3b2/lib/google/apis/core/upload.rb#L77
+      def prepare_io(io)
+        return io if streamable?(io)
+
+        if io.respond_to?(:to_io)
+          io.to_io
+        elsif io.respond_to?(:tempfile)
```

Comment:
Why do we need both `#to_io` and `#tempfile`? `ActionDispatch::Http::UploadedFile` responds to `#to_io` as well as `#tempfile` (they're aliases).

Reply:
In that case, what do you think about simplifying the whole method to

```rb
io.respond_to?(:to_io) ? io.to_io : io
```

Reply:
According to @herenow above:
> Older versions of Rails, such as Rails 3.2, doesn't yet have the
to_io method, in this case, we need to check for a tempfile.

Reply:
@janko-m Correct me if I'm wrong, but our `#upload` method receives a `ActionDispatch::Http::UploadedFile` and not a `Shrine::UploadedFile` right?

In this case, I included the `#tempfile` method because ActionDispatch::Http::UploadedFile prior to Rails 4.x doesn't respond to `#to_io`, but it does give me access to `#tempfile` [1].

I added it to support this older versions of Rails.

[1] https://github.com/rails/rails/blob/3-2-stable/actionpack/lib/action_dispatch/http/upload.rb#L4

Reply:
If I remember correctly, when I debugged the `#upload` method, `io` was a `ActionDispatch::Http::UploadedFile` instance. If it's is not, checking for tempfile is probably confusing and out of context.

Reply:
@herenow It can also receive a `Shrine::UploadedFile` object, when uploading a cached file unprocessed or when using the `restore_cahed_data` or `refresh_metadata` plugins. But since it also responds to `#to_io`, it's covered here.

Sorry, I missed the part about supporting older Rails versions, definitely agree on that 👍

But you might not need the `streamable?` check, as `#to_io` returns the receiver on all `IO` subclasses (this includes `File`), on `Tempfile` it returns the underlying `File` object, and on `StringIO` it doesn't exist. So it seems that all three classes are covered by the checks below.

Reply:
To clarify, there are a number of IO objects that can be given to `Storage#upload` depending on the context:

* `ActionDispatch::Http::UploadedFile`
* `Shrine::UploadedFile` (when uploading cached file)
* `Shrine::Plugins::RackFile::UploadedFile` (from `rack_file` plugin)
* `Shrine::Plugins::DataUri::DataFile` (from `data_uri` plugin)
* `Shrine::Plugins::RemoteUrl::RemoteFile` (from `remote_url` plugin, upcoming)
* `Tempfile` (from `remote_url` plugin)

All of them are covered by the checks you've provided.

Reply:
@janko-m I was hesitant to add the `streamable?` check, I added it just because `Tempfile#to_io` returns me a `File`, and gcs client doesn't accept `File` as a `upload_source`, since it is not streamable.

But checking for `IO` and `StringIO` is redundant, we should remove it, and just check if it is a tempfile.

Reply:
Actually, never mind, a `File` is an `IO`. I'll remove this.